### DISCUSSION TITLE
Squeeze-ESP32 PCB based on the TTGO WROVER v1.8

### DISCUSSION THREAD

**Discussion by zelea2** [sle118/squeezelite-esp32#266] [General]
I've made a schematic and PCB to build myself a few squeezelite devices to work with my Logitech media server.
I'm using the TTGO WROVER v1.8 module (price around 10$) because this already has the USB to serial chip on 
board and it's very convenient to program and debug. It also has a connector for a LiIon battery and the charging 
chip for it.
I don't want to have the audio power amplifier on this board because one can buy dozens of class D amplifier
boards very cheap if needed. Also it is possible to connect it to a sound bar (via analog stereo or SPDIF) and again
there are dozens of cheap and powerful sound bars out there.

I've chosen to use the ES9023 DAC chip (rather than PCM5102) because it is cheaper - you can buy 10pcs for 12$ from Aliexpress - and it has an easier to solder footprint than the PCM51xx. The drawback of using this chip is that it also
needs a master clock (which is provided by the ESP32) and doesn't have a PLL mode like the PCM51xx.

I've connected an SPI OLED display (with 3 superimposed footprints from monochrome 0.96" 7pins 128X64 to 2" 8pins 240x320 with ST7789).

Ethernet is provided by a W5500 module (just 4$) and it shares the SPI with the display. I'm using the HSPI (host=2) for this since it is not used by the WROVER module (Flash is SPI0 and PSRAM is SPI3/VSPI). 

I have an SPDIF optical connector (10pcs for 5$) driven directly by the ESP32.

I've also placed a footprint for a switch rotary encoder to navigate the screen display but if you solder the encoder then
you will lose the infrared receiver (VS1838) and the battery  voltage measurement (not enough free pins on the WROVER).

The PCB was targeted more to be easy to solder rather than smallest size so it is 75x95mm. The BOM is bellow 25$ (per board) for 10 boards (including the JLCPCB PCB and components from Aliexpress).

Any comments before I sent this PCB to manufacture?
[Schematic - Squeeze-ESP32.pdf](https://github.com/sle118/squeezelite-esp32/files/11882720/Squeeze-ESP32.pdf)
![Squeeze-ESP32-PCB](https://github.com/sle118/squeezelite-esp32/assets/48109409/fef7d51c-e7ca-4198-9cb3-e2007ab8100b)
[PCB - Squeeze-ESP32-PCB.pdf](https://github.com/sle118/squeezelite-esp32/files/11882723/Squeeze-ESP32-PCB.pdf)

**Comment by sle118** [comment]
A few things here. First, you should leave some room around the board and add some holes to help with fixing the board in, say, a printed enclosure. You ideally also want to have all the connectors editing from the same side so an enclosure would look good in a typical setup, with all the wires sticking out from the back. Although it is hard for me to really see what you've done here

Which software did you use to design the board? Does it have a 3d rendering capability? Without a schematic or rendering, it is difficult to give more feedback.

You could also describe the features in more details, as it will also help in understanding your state of mind when you made some design decisions. For example, easy to solder is a start.

**Comment by zelea2** [comment]
The schematic was linked above but I'll include it again here.
I've found some plastic enclosures from China (some 3$ shipped) which are 125x80x32 mm (model AK-S-30) and I've made the
PCB to fit inside this. At 32mm it is quite thick so there is plenty of space to include a LiIon battery for autonomous operation.
I'm using an old OrCAD version and I don't have 3D models for my components.
[Squeeze-ESP32.pdf](https://github.com/sle118/squeezelite-esp32/files/11910303/Squeeze-ESP32.pdf)
![Squeeze-ESP32](https://github.com/sle118/squeezelite-esp32/assets/48109409/f1d9bc2f-afd7-45dc-aa70-bf15d10de08f)
![Squeeze-ESP32](https://github.com/sle118/squeezelite-esp32/assets/48109409/89372755-ce5f-44d1-bb4d-dd35b1a41852)
![Squeeze-ESP32-PCB](https://github.com/sle118/squeezelite-esp32/assets/48109409/5519f1d1-4e75-4d1a-b0d3-9bf3b478e78c)
![box_125x80x32](https://github.com/sle118/squeezelite-esp32/assets/48109409/9697c4e7-a71d-45e8-8c49-1f6e0597dcc2)

**Comment by sle118** [comment]
From there, all I have to say is the proof of the pudding is in the tasting 😁

**Reply by zelea2** [reply]
These are the relevant configuration entries for the above board:

- CONFIG_DAC_CONFIG="model=I2S,bck=33,ws=25,do=32,mck=0,mute=22:0"
- CONFIG_SPDIF_CONFIG="bck=33,ws=25,do=5"
- CONFIG_SPI_CONFIG="mosi=13,miso=12,clk=14,host=2,dc=26"
- CONFIG_DISPLAY_CONFIG="width=128,height=128,speed=15000000,HFlip,VFlip,driver=SSD1306,cs=2
- CONFIG_ETH_CONFIG="model=w5500,speed=40000000,cs=15,intr=21,rst=4,host=-1"
- CONFIG_BAT_CONFIG="channel=7,scale=7.11,atten=3,cells=1"
- CONFIG_SET_GPIO="12=red:0,13=green:0,36=ir,34=jack:0"
- CONFIG_ROTARY_ENCODER="A=36,B=39,SW=35"

The display entry should be adjusted depending on which of the 3 types of OLED is used.

While browsing the code I've just noticed that the encoder cannot use IOs 36&39 because the ESP32 has a bug and 
generates continuous interrupts. So the rotary encoder doesn't seem to be usable with the above schematic and
I've already sent the boards to manufacture (and no free pins anyways). The IR seems to be fine with IO36 though 
(RMT peripheral).
I might change the rotary encoder code to use polling instead of interrupts since the overhead is not big.

**Reply by sle118** [reply]
Ah yes I forgot about the gpios issue, and that wasn't obvious either from looking at the board layout. You can always cut traces and reroute pins with small wires.

**Reply by zelea2** [reply]
Don't need to cut any traces since there are no free pins left. If I don't use the ethernet  module then I'll solder 2 wires from the encoder to Eth_nRst/nInt pins.

**Reply by sle118** [reply]
In a future iteration, you could also consider adding one of the many io expansion modules available. It is possible to share the communication bus with Ethernet on some as well: https://github.com/sle118/squeezelite-esp32#gpio-expanders

That could very well be soldered on the back of the board to keep it the same size.

**Reply by zelea2** [reply]
I've already considered it, even made a footprint for the PCA8575 a module, but decided against in the end. Those I2C expanders are very slow, the bus is maxed at 400KHz so with 16 bits per transfer that already 40us and if you add the time
to process the interrupt you're lucky if you can read pins at 10KHz sampling rate and that's only good for LEDs, buttons and encoders. What's worst is that it occupies 3 more pins: SCL, SDA and Int (I'm currently not using the I2C bus at all).

**Reply by sle118** [reply]
In my own designed board,  I have display,  expander and Ethernet on the same spi bus and they go along really well. I went for the MCP23S17, which is capable and provides plenty of pins.

**Reply by zelea2** [reply]
SPI expanders are nice and operate at a much higher speed but right now it is difficult to get already built modules cheap. Aliexpress has plenty of MCP23017 ones but those are I2C. Maybe next revision as you said.

**Reply by sle118** [reply]
@zelea2 make sure you post some pictures here when you have assembled one or more boards!

Don't forget here as well https://forums.slimdevices.com/forum/user-forums/3rd-party-hardware/111066-share-your-self-made-squeezelite-esp32-hardware-project

**Comment by sle118** [comment]
@zelea2 I am curious if you made some progress with your project.
